site/src/pages/TemplateVersionEditorPage/TemplateVersionEditorPage.tsx

```diff
@@ -1,115 +1,359 @@
+    enabled: templateVersion !== undefined,
+  });
+  const [fileTree, setFileTree] = useState<FileTree>();
+  useEffect(() => {
+    const initializeFileTree = async (file: ArrayBuffer) => {
+      const tarReader = new TarReader();
+      await tarReader.readFile(file);
+      tarFileRef.current = tarReader;
+      const fileTree = await createTemplateVersionFileTree(tarReader);
+      setFileTree(fileTree);
+    };
+
+    if (fileQuery.data) {
+      initializeFileTree(fileQuery.data).catch(() => {
+        console.error("Error on initializing the editor");
+      });
+    }
+  }, [fileQuery.data]);
+
+  return {
+    fileTree,
+    tarFile: tarFileRef.current,
+  };
+};
+
+const useVersionLogs = (
+  templateVersion: TemplateVersion | undefined,
+  options: { onDone: () => Promise<unknown> },
+) => {
+  const [logs, setLogs] = useState<ProvisionerJobLog[]>();
+  const templateVersionId = templateVersion?.id;
+  const refetchTemplateVersion = options.onDone;
+  const templateVersionStatus = templateVersion?.job.status;
+
+  useEffect(() => {
+    if (!templateVersionId || !templateVersionStatus) {
+      return;
+    }
+
+    if (templateVersionStatus !== "running") {
+      return;
+    }
+
+    const socket = watchBuildLogsByTemplateVersionId(templateVersionId, {
+      onMessage: (log) => {
+        setLogs((logs) => (logs ? [...logs, log] : [log]));
+      },
+      onDone: async () => {
+        await refetchTemplateVersion();
+      },
+      onError: (error) => {
+        console.error(error);
+      },
+    });
+
+    return () => {
+      socket.close();
+    };
+  }, [refetchTemplateVersion, templateVersionId, templateVersionStatus]);
+
+  return {
+    logs,
+    setLogs,
+  };
+};
+
+const useMissingVariables = (templateVersion: TemplateVersion | undefined) => {
+  const { data: missingVariables } = useQuery({
+    ...templateVersionVariables(templateVersion?.id ?? ""),
+    enabled: templateVersion?.job.error_code === "REQUIRED_TEMPLATE_VARIABLES",
+  });
+  const [isMissingVariablesDialogOpen, setIsMissingVariablesDialogOpen] =
+    useState(false);
+
+  useEffect(() => {
+    if (missingVariables) {
+      setIsMissingVariablesDialogOpen(true);
+    }
+  }, [missingVariables]);
+
+  return {
+    missingVariables,
+    isMissingVariablesDialogOpen,
+    setIsMissingVariablesDialogOpen,
+  };
+};
+
+const generateVersionFiles = async (
+  tarReader: TarReader,
+  fileTree: FileTree,
+) => {
+  const tar = new TarWriter();
+
+  // Add previous non editable files
+  for (const file of tarReader.fileInfo) {
+    if (!isAllowedFile(file.name)) {
+      if (file.type === "5") {
+        tar.addFolder(file.name, {
+          mode: file.mode, // https://github.com/beatgammit/tar-js/blob/master/lib/tar.js#L42
+          mtime: file.mtime,
+          user: file.user,
+          group: file.group,
+        });
+      } else {
+        tar.addFile(file.name, tarReader.getTextFile(file.name) as string, {
```

Comment:
Should there be some error handling in the off chance `tarReader.getTextFile` returns `undefined`?

Reply:
For this specific use case, we can just ignore it for now.